general/mem_manager.hpp

```diff
@@ -26,28 +26,64 @@ namespace mfem
 /// Memory types supported by MFEM.
 enum class MemoryType
 {
-   HOST,      ///< Host memory; using new[] and delete[]
-   HOST_32,   ///< Host memory aligned at 32 bytes (not supported yet)
-   HOST_64,   ///< Host memory aligned at 64 bytes (not supported yet)
-   CUDA,      ///< cudaMalloc, cudaFree
-   CUDA_UVM   ///< cudaMallocManaged, cudaFree (not supported yet)
+   HOST,           ///< Host memory; using new[] and delete[]
+   HOST_32,        ///< Host memory; aligned at 32 bytes
+   HOST_64,        ///< Host memory; aligned at 64 bytes
+   HOST_DEBUG,     ///< Host memory; protected while in device mode
```

Comment:
What does "device mode" mean here?

Is it correct to say that this type is `Host memory allocated from a "host-debug" pool` as opposed to `DEVICE_DEBUG` which is `Pseudo-device memory allocated on host from a "device-debug" pool`? A pool here mean a section of host memory that can be protected against writing when switch between "host mode" and "device mode".

Reply:
Updated comment.

Reply:
With the first question, I was wondering how the protection worked, trying to understand the comment "protected while in device mode". After looking at the implementation, I think I now understand how it works: if you request the host pointer from a `Memory` object, the host pointer is unprotected and the device is protected; and vice-versa. Is this correct?

Also, I'm wondering if the protection can lead to false positives, i.e. generate errors when the code is actually OK. For example, if the `Memory` is valid both on host and device and the device is un-protected (host is protected); then if we try to read the host memory using the pointer `Memory::h_ptr` (e.g. using `operator[](int) const`) I'd expect to get an error because the host-pointer is protected. Is this scenario possible?

Reply:
You're correct and the protections should be set/unset regarding the requested access.
I'll take a look at the case you're describing, but it shouldn't raise any false positive. If both are valid, both should be accessible.
The main difficulty is to handle aliases: sometimes, they are too small to allow one page to be  protected, so it won't trigger any protection.

Reply:
Ideally, the protections will be sync'd with the validity flags: if both are valid -- only read access is allowed to both pointers; if only one is valid -- the other is protected against read and write. This, however, will probably require more work which we don't need to do right now.